File: server/php/src/Controller/ServerController.php
<?php

namespace App\Controller;

use Mustache_Engine;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\JsonResponse;
use Braintree\Gateway;

class ServerController extends AbstractController
{
    private $mustache;

    public function __construct()
    {
        $this->mustache = new Mustache_Engine(["entity_flags" => ENT_QUOTES]);
    }

    /* ######################################################################
     * Token generation helpers
     * ###################################################################### */

    public function getClientToken()
    {
        $gateway = BraintreeGateway::getGateway();

        $clientToken = $gateway->clientToken()->generate();

        return new JsonResponse(["clientToken" => $clientToken]);
    }

    /* ######################################################################
     * Serve checkout page
     * ###################################################################### */

    public function index(Request $request)
    {
        $isFlexibleIntegration = $request->query->get("flexible", false);

        $locals = [
            "title" =>
                "Fastlane - Braintree SDK Integration" .
                ($isFlexibleIntegration ? " (Flexible)" : ""),
            "prerequisiteScripts" => '
                <script
                    src="https://js.braintreegateway.com/web/3.106.0/js/client.min.js"
                    defer
                ></script>
                <script
                    src="https://js.braintreegateway.com/web/3.106.0/js/data-collector.min.js"
                    defer
                ></script>
                <script
                    src="https://js.braintreegateway.com/web/3.106.0/js/fastlane.min.js"
                    defer
                ></script>
            ',
            "initScriptPath" => $isFlexibleIntegration
                ? "init-fastlane-flexible.js"
                : "init-fastlane.js",
            "stylesheetPath" => "styles.css",
        ];

        $htmlTemplate = file_get_contents(
            __DIR__ .
                "/../../../shared/views/" .
                ($isFlexibleIntegration
                    ? "checkout-flexible.html"
                    : "checkout.html")
        );

        $template = $this->mustache->render($htmlTemplate, $locals);

        return new Response($template);
    }

    /* ######################################################################
     * Process transactions
     * ###################################################################### */

    public function createTransaction(Request $request)
    {
        $data = $request->toArray();

        $gateway = BraintreeGateway::getGateway();

        $result = $gateway->transaction()->sale([
            "amount" => "10.00",
            "paymentMethodNonce" => $data["paymentToken"]["id"],
            "deviceData" => $data["deviceData"],
            "options" => [
                "submitForSettlement" => true,
            ],
        ]);

        return new JsonResponse(["result" => $result]);
    }
}

class BraintreeGateway
{
    private static $gateway = null;

    public static function getGateway()
    {
        if (self::$gateway === null) {
            self::$gateway = new Gateway([
                "environment" => "sandbox",
                "merchantId" => Env::get("BRAINTREE_MERCHANT_ID"),
                "publicKey" => Env::get("BRAINTREE_PUBLIC_KEY"),
                "privateKey" => Env::get("BRAINTREE_PRIVATE_KEY"),
            ]);
        }

        return self::$gateway;
    }
}

class Env
{
    static function get(string $key): string|null
    {
        return $_ENV[$key] ?? null;
    }
}

File: server/java/src/main/java/com/fastlane/braintreesdksample/sample/ServerController.java
package com.fastlane.braintreesdksample.sample;

import com.braintreegateway.BraintreeGateway;
import com.braintreegateway.ClientTokenRequest;
import com.braintreegateway.Environment;
import com.braintreegateway.Result;
import com.braintreegateway.Transaction;
import com.braintreegateway.TransactionAddressRequest.ShippingMethod;
import com.braintreegateway.TransactionRequest;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fastlane.braintreesdksample.sample.models.ClientTokenResponse;
import com.fastlane.braintreesdksample.sample.models.CreateTransactionRequest;
import io.github.cdimascio.dotenv.Dotenv;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.servlet.ModelAndView;

@RestController
public class ServerController {

    private final String title = "Fastlane - Braintree SDK Integration";
    private final String prerequisiteScripts =
        """
        <script src=\"https://js.braintreegateway.com/web/3.106.0/js/client.min.js\" defer></script>
        <script src=\"https://js.braintreegateway.com/web/3.106.0/js/data-collector.min.js\" defer ></script>
        <script src=\"https://js.braintreegateway.com/web/3.106.0/js/fastlane.min.js\" defer ></script>
        """;
    private final String initScriptPath = "init-fastlane%s.js";
    private final String stylesheetPath = "../../styles.css";

    private final Dotenv dotenv;
    private final BraintreeGateway gateway;
    private final ArrayList<String> domains;

    public ServerController() {
        this.dotenv = Dotenv.load();
        this.gateway = new BraintreeGateway(
            Environment.SANDBOX,
            dotenv.get("BRAINTREE_MERCHANT_ID"),
            dotenv.get("BRAINTREE_PUBLIC_KEY"),
            dotenv.get("BRAINTREE_PRIVATE_KEY")
        );
        this.domains = new ArrayList<>(Arrays.asList(dotenv.get("DOMAINS").split(",")));
    }

    /* ######################################################################
     * Token generation helpers
     * ###################################################################### */

    @CrossOrigin
    @GetMapping("/client-token")
    public ResponseEntity<?> getClientToken() {
        try {
            ClientTokenRequest clientTokenRequest = new ClientTokenRequest().domains(domains);

            String clientToken = gateway.clientToken().generate(clientTokenRequest);

            return new ResponseEntity<>(new ClientTokenResponse(clientToken), HttpStatus.OK);
        } catch (Exception e) {
            return new ResponseEntity<>(e.getMessage(), HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }

    /* ######################################################################
     * Serve checkout page
     * ###################################################################### */

    @GetMapping("/")
    public ModelAndView getCheckout(@RequestParam(name = "flexible", required = false) String isFlexible, Model model) {
        model.addAttribute("title", isFlexible != null ? title + " (Flexible)" : title);
        model.addAttribute("prerequisiteScripts", prerequisiteScripts);
        model.addAttribute(
            "initScriptPath",
            isFlexible != null ? String.format(initScriptPath, "-flexible") : String.format(initScriptPath, "")
        );
        model.addAttribute("stylesheetPath", stylesheetPath);

        String page = isFlexible != null ? "checkout-flexible" : "checkout";

        return new ModelAndView(page, model.asMap());
    }

    /* ######################################################################
     * Process transactions
     * ###################################################################### */

    @CrossOrigin
    @PostMapping("/transaction")
    public ResponseEntity<?> createTransaction(@RequestBody CreateTransactionRequest body) {
        try {
            ObjectMapper mapper = new ObjectMapper();
            JsonNode rootNode = mapper.valueToTree(body);
            JsonNode nameNode = rootNode.at("/name");
            JsonNode billingAdressNode = rootNode.at("/paymentToken/paymentSource/card/billingAddress");
            JsonNode shippingAdressNode = rootNode.at("/shippingAddress");

            TransactionRequest request = new TransactionRequest()
                .amount(new BigDecimal(10))
                .paymentMethodNonce(body.getPaymentToken().getId())
                .deviceData(body.getDeviceData())
                .customer()
                .email(rootNode.path("email").textValue())
                .firstName(nameNode.path("firstName").textValue())
                .lastName(nameNode.path("lastName").textValue())
                .done()
                .billingAddress()
                .streetAddress(billingAdressNode.path("streetAddress").textValue())
                .extendedAddress(billingAdressNode.path("extendedAddress").textValue())
                .locality(billingAdressNode.path("locality").textValue())
                .region(billingAdressNode.path("region").textValue())
                .postalCode(billingAdressNode.path("postalCode").textValue())
                .countryCodeAlpha2(billingAdressNode.path("countryCodeAlpha2").textValue())
                .done()
                .shippingAddress()
                .streetAddress(shippingAdressNode.path("streetAddress").textValue())
                .extendedAddress(shippingAdressNode.path("extendedAddress").textValue())
                .locality(shippingAdressNode.path("locality").textValue())
                .region(shippingAdressNode.path("region").textValue())
                .postalCode(shippingAdressNode.path("postalCode").textValue())
                .countryCodeAlpha2(shippingAdressNode.path("countryCodeAlpha2").textValue())
                .shippingMethod(ShippingMethod.GROUND)
                .done()
                .options()
                .submitForSettlement(true)
                .done();

            Result<Transaction> result = gateway.transaction().sale(request);

            if (!result.isSuccess()) {
                return new ResponseEntity<>(result.getMessage(), HttpStatus.BAD_REQUEST);
            }

            Map<String, Object> response = new HashMap<>();
            Map<String, Object> resultResponse = new HashMap<>();
            resultResponse.put("success", true);
            resultResponse.put("transaction", result.getTarget());
            response.put("result", resultResponse);

            return new ResponseEntity<>(response, HttpStatus.OK);
        } catch (Exception e) {
            return new ResponseEntity<>(e.getMessage(), HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }
}
